### PR TITLE
[FEAT] Implements Decorators

### PR DESCRIPTION
**Pull request by pzuraq**
This is a continuation of #17226, thanks a ton to @NullVoxPopuli for getting this started!

This PR refactors the Descriptor system in Ember metal, mostly replacing
it with a Decorator system instead. `defineProperty` has been updated to
apply decorators following the currently implemented spec. This means
that the value returned by `computed` goes through the exact same flow
when applied as a decorator using decorator syntax, and applied using
classic syntax.

I was considering pulling out the ComputedDecorator stuff into a separate
package, it's relatively isolated from the rest of `metal` at this point
so it would make some sense. Thoughts?

TODO:
- [x] Benchmark
- [x] Disable native application except behind a canary feature flag

### CODE REVIEW

File: broccoli/packages.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ const WriteFile = require('broccoli-file-creator');
 const StringReplace = require('broccoli-string-replace');
 const GlimmerTemplatePrecompiler = require('./glimmer-template-compiler');
 const VERSION_PLACEHOLDER = /VERSION_STRING_PLACEHOLDER/g;
+const transfromBabelPlugins = require('./transforms/transform-babel-plugins');
```
Comment (richard-viney) [inline broccoli/packages.js]:
```suggestion
const transformBabelPlugins = require('./transforms/transform-babel-plugins');
```

File: broccoli/packages.js [hunk 2]
```diff
@@ -83,6 +84,13 @@ module.exports.getPackagesES = function getPackagesES() {
     exclude: ['**/*.ts'],
   });
 
+  // tsc / typescript handles decorators and class properties on its own
+  // so for non ts, transpile the proposal features (decorators, etc)
+  let transpiledProposals = debugTree(
+    transfromBabelPlugins(debugTree(nonTypeScriptContents, `get-packages-es:babel-plugins:input`)),
```
Comment (richard-viney) [inline broccoli/packages.js]:
```suggestion
    transformBabelPlugins(debugTree(nonTypeScriptContents, `get-packages-es:babel-plugins:input`)),
```

**Comment by pzuraq** [comment]
Thanks @richard-viney!

File: packages/@ember/-internals/metal/lib/computed.ts [hunk 3]
```diff
@@ -567,6 +600,7 @@ if (EMBER_METAL_TRACKED_PROPERTIES) {
 
   ```js
   import EmberObject, { computed } from '@ember/object';
+import { DecoratorDescriptor } from '../../../../../tmp/funnel-input_base_path-gav85kkw.tmp/@ember/-internals/metal/lib/decorator_descriptor';
```
Comment (ro0gr) [inline packages/@ember/-internals/metal/lib/computed.ts]: Something wrong with import path

Comment (pzuraq) [reply packages/@ember/-internals/metal/lib/computed.ts]: Ah, automated refactor by VSCode, will fix

File: packages/@ember/-internals/metal/lib/alias.ts [hunk 4]
```diff
@@ -94,7 +118,3 @@ function AliasedProperty_oneWaySet(obj: object, keyName: string, value: any): an
   defineProperty(obj, keyName, null);
   return set(obj, keyName, value);
 }
-
-// Backwards compatibility with Ember Data.
-(AliasedProperty.prototype as any)._meta = undefined;
-(AliasedProperty.prototype as any).meta = ComputedProperty.prototype.meta;
```
[the remark this answers is not in the document]
Comment (pzuraq) [reply packages/@ember/-internals/metal/lib/alias.ts]: These still exist on the property/prototype, I just moved them from here. `meta` is still usable on it.

**Comment by pzuraq** [comment]
Benchmarked, no statistically significant change in performance! I think this is ready for review 🎉 

[results.json.txt](https://github.com/emberjs/ember.js/files/2830618/results.json.txt)
[results.pdf](https://github.com/emberjs/ember.js/files/2830617/results.pdf)

File: packages/@ember/-internals/metal/lib/alias.ts [hunk 5]
```diff
@@ -74,14 +100,12 @@ export class AliasedProperty extends Descriptor implements DescriptorWithDepende
     return set(obj, this.altKey, value);
   }
 
-  readOnly(): this {
+  readOnly(): void {
```
[the remark this answers is not in the document]
Comment (pzuraq) [reply packages/@ember/-internals/metal/lib/alias.ts]: These are chainable. The public value is now a function, and it's prototype is set to `AliasDecoratorImpl`: https://github.com/emberjs/ember.js/pull/17548/files/4359400a8d6fe78708f5fe0cb5b941d2a5dc3eda#diff-7945edc029962d6f791f27bd1eb5d30cR26

This is what users actually end up interacting with. Internally, it calls these functions on the descriptor, which is now an entirely private implementation detail within Ember.

File: packages/@ember/-internals/metal/lib/alias.ts [hunk 6]
```diff
@@ -2,25 +2,50 @@ import { Meta, meta as metaFor } from '@ember/-internals/meta';
...
 
-export default function alias(altKey: string): AliasedProperty {
-  return new AliasedProperty(altKey);
+export type AliasDecorator = Decorator & PropertyDecorator & AliasDecoratorImpl;
+
+export default function alias(altKey: string): AliasDecorator {
+  return makeComputedDecorator(new AliasedProperty(altKey), AliasDecoratorImpl) as AliasDecorator;
+}
+
+class AliasDecoratorImpl {
```
Comment (pzuraq) [inline packages/@ember/-internals/metal/lib/alias.ts]: This should extend `Function`

Comment (pzuraq) [reply packages/@ember/-internals/metal/lib/alias.ts]: Also, add a comment to svelte this once `computed().meta()` has been deprecated

File: packages/@ember/-internals/views/lib/mixins/view_support.js [hunk 7]
```diff
@@ -143,11 +142,11 @@ let mixin = {
   /**
    Returns the current DOM element for the view.
 
-   @property element
-   @type DOMElement
-   @public
-   */
-  element: descriptor({
```
Comment (pzuraq) [inline packages/@ember/-internals/views/lib/mixins/view_support.js]: Followup: We should flatten this class into `Component` so we don't need to use this decorator

File: packages/ember/index.js [hunk 8]
```diff
@@ -271,8 +271,9 @@ Object.defineProperty(Ember.run, 'currentRunLoop', {
 const computed = metal._globalsComputed;
 Ember.computed = computed;
 computed.alias = metal.alias;
-Ember.ComputedProperty = metal.ComputedProperty;
 Ember.cacheFor = metal.getCachedValueFor;
+Ember.ComputedProperty = metal.ComputedProperty;
+Ember.setComputedDecorator = metal.setComputedDecorator;
```
Comment (pzuraq) [inline packages/ember/index.js]: Underscore this to imply privacy

File: packages/@ember/-internals/metal/lib/decorator.ts [hunk 9]
```diff
@@ -0,0 +1,240 @@
+import { Meta, meta as metaFor, peekMeta } from '@ember/-internals/meta';
+import { EMBER_NATIVE_DECORATOR_SUPPORT } from '@ember/canary-features';
+import { assert } from '@ember/debug';
+import { DEBUG } from '@glimmer/env';
+import { unwatch, watch } from './watching';
+
+const DECORATOR_DESCRIPTOR_MAP: WeakMap<Decorator, any> = new WeakMap();
```
Comment (pzuraq) [inline packages/@ember/-internals/metal/lib/decorator.ts]: `boolean | ComputedDescriptor`

File: packages/@ember/-internals/metal/lib/decorator.ts [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+/**
+  Check whether a value is a decorator
+
+  @method isComputedDecorator
+  @param {any} possibleDesc the value to check
+  @return {boolean}
+  @private
+*/
+export function isComputedDecorator(dec: Decorator | null | undefined) {
+  return dec === null || dec === undefined ? false : DECORATOR_DESCRIPTOR_MAP.has(dec);
```
Comment (pzuraq) [inline packages/@ember/-internals/metal/lib/decorator.ts]: `&&`

File: packages/@ember/-internals/metal/lib/computed.ts [hunk 11]
```diff
@@ -153,77 +150,131 @@ function noop(): void {}
...
+  private _suspended: any = undefined;
+  private _hasConfig = false;
+
+  _getter?: ComputedPropertyGetter = undefined;
+  _setter?: ComputedPropertySetter = undefined;
   _auto?: boolean;
-  _dependentKeys: string[] | undefined;
 
-  constructor(config: ComputedPropertyConfig, opts?: ComputedPropertyOptions) {
+  constructor(args: Array<string | ComputedPropertyConfig>) {
```
Comment (buschtoens) [inline packages/@ember/-internals/metal/lib/computed.ts]: Was this change intentional? Shouldn't it be `...args` instead?

Comment (buschtoens) [reply packages/@ember/-internals/metal/lib/computed.ts]: Confirmed on Discord that it is intentional.
